Merge TigerBeetle clients into main repo

cb22 opened the pull request:
This PR pulls in tigerbeetle-{node,go,java,dotnet}, as well as their respective CI workflows, and adds some extra tooling around publishing to npm and the Go client. 

All commits in this PR - except for the single commit titled `clients: Import clients and scope CI workflows into main repo` - are said imports and can mostly be skipped for review purposes.

- All client commits have been preserved and rewritten to start with `clients/{client}: `. These clients now live under the `src/clients/` subfolder
- Binaries have been removed from the history of the Go and Node clients

For Go:
- To keep binaries out of the main repo, the Go client workflow builds the binaries needed and then pushes the result to the `tigerbeetle-go` repo. It includes the source commit in its commit message, as well as creates a tag called `tigerbeetle-{commit-in-this-repo}` for easy reference.

For Node:
- When working on a PR, the Node workflow will check if the version specified in `package.json` exists on npm. If it does, the hash of the npm hosted package is then compared to a locally built one. 
    - It would be nice to only include / rely on the files that the client _actually_ uses and builds in. At the moment, the detection for this will likely have some false positives.
- If the hashes aren't identical, it'll throw an error. The idea is to stop cases where changes have been made to the client, but a new release hasn't been cut (thanks @sentientwaffle!)
- When pushing to main or merging, a new release can be created by simply adjusting the version number in `package.json`

Before merging:
- [x] The `tigerbeetle-go` repo needs to have a deploy key with RW access added, and that key added as a secret here under `TIGERBEETLE_GO_DEPLOY_KEY`
- [x] A key for npm publishing needs to be added as a secret under `TIGERBEETLE_NODE_PUBLISH_KEY`

Other points:
- Currently, the workflows for the clients are not scoped to the client path; they'll trigger if something unrelated is changed. While changing the `dotnet` client will very likely have 0 bearing on the `java`, this is useful to catch modifications and build for things like the core Zig client library that they pull in.

sublimator committed:
clients/node: Reduce log level to err

kprotty committed:
clients/go: Merge pull request #2 from coilhq/main-sync

Update to use tigerbeetle/main

barnarddt committed:
clients/go: Add type references to the Event results for Accounts and Transfers
Add stringger code for the resutlts

This will allow fmt and log packages to print error messages instead of uint32 that they would need to look at the source to understand what the code means. This is a developer ease change and not a functional change.

barnarddt committed:
clients/go: Second thoughts, lets leave the Valididator for now

barnarddt committed:
clients/go: Changes from review

DonChangfoot committed:
clients/go: Merge pull request #4 from barnarddt/main

Returned types instead of raw error codes Accounts and Transfers

barnarddt committed:
clients/go: Regenerate stringer functions

matdehaast committed:
clients/go: Merge pull request #5 from barnarddt/main

Regenerate stringer functions

barnarddt committed:
clients/go: * Update enum names to conform with tigerbeetle zig naming
* Fix unit tests

matdehaast committed:
clients/go: Merge pull request #7 from barnarddt/main

Fix unit testing for latest version of tigerbeetle

barnarddt committed:
clients/go: fix(types): regenerate string functions for account and transfer results

koekiebox committed:
clients/node: Changes include:
* Initial changes for 'lsm' fixes.
* Updated error codes for create account and transfer.

koekiebox committed:
clients/node: Changes include:
* Initial changes for 'lsm' fixes.
* Updated error codes for create account and transfer.

koekiebox committed:
clients/node: Changes include:
* Updated command for tigerbeetle benchmark.sh

sentientwaffle committed:
clients/node: node.zig: Update for MessagePool refactor

`MessagePool` is owned by the `Context`, so it must not be allocated on the stack.

Commit:
clients/go: Update go client for tigerbeetle lsm branch

Commit:
clients/go: Update git submodule to point at tigerbeetle lsm branch

koekiebox committed:
clients/node: Changes include:
* README.md - Update for latest commands.

koekiebox committed:
clients/node: Changes include:
* Bump submodule file.

koekiebox committed:
clients/node: Changes include:
* README.md - Updated commands.

koekiebox committed:
clients/node: Changes include:
* README.md - Updated commands.

koekiebox committed:
clients/node: Changes include:
* README.md - Updated commands.

jorangreef committed:
clients/node: Merge pull request #27 from coilhq/fixes/tb-lsm-fixes

fix(lsm): Fixes for TigerBeetle LSM updates

jorangreef committed:
clients/node: 0.10.0

Commit:
clients/go: Update submodule to main since lsm was merged

kprotty committed:
clients/go: Merge pull request #9 from coilhq/lsm-sync

fix(lsm): Update to support tigertbeetle lsm branch

batiati committed:
clients/dotnet: Initial commit

This project is a C# implementation of the official TigerBeetle client API "tb_client"

Features

- Dotnet 6.0+ and dotnet standard 2.1 support
- Async and blocking APIs
- Idiomatic C#/dotnet naming conventions
- Full featured client
- Basic integration tests with a live TigerBeetle server

Pending

- Build system with the unmanaged library as a embedded resource
- More tests, and stress tests for the PacketList syncronization

batiati committed:
clients/dotnet: ReadMe and License

batiati committed:
clients/dotnet: More tests

- Integration tests VoidPendingTransfers and LinkedTransfers

batiati committed:
clients/dotnet: Code review

- Remove unecessary unsafe decls
- Add readonly modifier when its possible
- Remove debug bits left in the code

batiati committed:
clients/dotnet: Apply dotnet format

batiati committed:
clients/dotnet: Tests running on linux

- Add libtb_client.so to the project
- Adjust the IntegrationTest helper to start and cleanup properly

batiati committed:
clients/dotnet: Add submodule

batiati committed:
clients/dotnet: Build with native libraries

- Add native libraries to windows, linux and macos x64

batiati committed:
clients/dotnet: ReadMe

- Add build instructions

batiati committed:
clients/dotnet: Update submodule

batiati committed:
clients/dotnet: Update submodule

batiati committed:
clients/dotnet: New build.zig script

batiati committed:
clients/dotnet: MSBuild updated

- Builds the tb_client from the MSBuild file
- Updates the ReadMe file

batiati committed:
clients/dotnet: Constructor refactory

- Idiomatic and simpler constructor

batiati committed:
clients/dotnet: Dotnet format

batiati committed:
clients/dotnet: ReadMe

- Add usage and basic API coverage

batiati committed:
clients/dotnet: Build support for Windows

batiati committed:
clients/dotnet: ReadMe

- Update build instructions for Windows
TODO: Add a `install.bat`

batiati committed:
clients/dotnet: Update TigerBeetle

- Update tigerbeetle submodule
- Update binaries
- Simplify the csproj prebuild

batiati committed:
clients/dotnet: Update submodule

batiati committed:
clients/dotnet: Fix csproj build on Windows

batiati committed:
clients/dotnet: Code review and cleanup

- Async review
- Missing StructLayout attribute
- dotnet format

batiati committed:
clients/java: Initial commit

- basic dir structure
- add jui submodule

batiati committed:
clients/java: Initial setup

- Java project setup
- VS Code config

batiati committed:
clients/dotnet: UInt128 review

batiati committed:
clients/dotnet: UInt128 review

- Add const types LU for unsigned longs
- unchecked cast when converting from signed integers (int and long)

batiati committed:
clients/dotnet: Build scripts review

- Rename "dynamic library" to "shared library"
- Use "addSharedLibrary" on build.zig

batiati committed:
clients/java: JUI submodule update

batiati committed:
clients/java: Build setup

batiati committed:
clients/java: Initial Java project

- Basic project structure
- Setup tests
- Batch memory model prototype

jorangreef committed:
clients/node: Merge pull request #24 from coilhq/nd-reduce-noise-2022-04-29

 Reduce log level to err

batiati committed:
clients/java: JNI improvements

- Batch memory model
- Zig JNI side

The java client can start and stop the native tb client through JNI
Next step, implement the batch and callback mechanism

batiati committed:
clients/java: Update submodule

batiati committed:
clients/java: Submodule update

batiati committed:
clients/java: More initial work

- CreateAccount works
- Initial memory and thread model are defined

Before implementing other operations:

- PacketList handling on Java or JNI client side
- Insert basic tests
- Reserach and review some best practices on JNI regarding references and syncronization
- Better Zig side code organization

batiati committed:
clients/java: Code refactory on Zig side

- Better organization
- JNI calls placed on "Reflection" structs
- Less glogal values

Still needs rework:

- Get rid of catch unreachable
- Review points where "wrapErrors" is used
- Fix "fromJString" function error handling

batiati committed:
clients/java: Lookup account

batiati committed:
clients/java: Initial protocol implementation

- All operations implemented (CreateAccounts,LookupAccounts,CreateTransfers,LookupTransfers)
- Basic benchmark

Needs work:

- Handle the packet list
- Tests

batiati committed:
clients/java: ReadMe

DonChangfoot committed:
clients/go: Merge pull request #8 from barnarddt/main

fix(types): regenerate string functions for account and transfer results

batiati committed:
clients/java: Future<T> support

- Added the singature Future<T> doAsync for all operations
- The Request class now implements a Future<T>, so the wait process can be done in a idiomatic way
- Exception handling for the Request

batiati committed:
clients/java: Code cleanup

- Java fmt
- drop local reference on zig side

batiati committed:
clients/java: Syncronization

- Added adquirePacket and returnPacket methods to manage the packet list
- Java side uses a semaphore to assure the max concurrent packets at once
- Overeall refactory: smaller syncronized blocks, fewer supressWarnings

Needs work:

- Remove zig side unreachable and panics
- Unit tests
- Integrated tests
- Multiplatform build and jar packages

lewisdaly committed:
clients/node: chore(readme): fix `createTransfers` example

jorangreef committed:
clients/node: Update README.md

jorangreef committed:
clients/node: Merge pull request #29 from lewisdaly/chore/28-readme-transfer-example

chore(readme): fix `createTransfers` example

barnarddt committed:
clients/go: fix(types): regenerate string functions for account and transfer results

DonChangfoot committed:
clients/go: Merge pull request #10 from barnarddt/main

fix(types): regenerate string functions for account and transfer results

batiati committed:
clients/java: Code refactory

- Move the serialization from the batch to the struct
This way we bypass get/set acessors without breaking java's idiomatic convention

- Rename some methods starting with capital letter

batiati committed:
clients/java: Submodule update

batiati committed:
clients/java: Submodule update

batiati committed:
clients/java: Build and bench scripts

- Added install.sh and benchmark.sh
- ReadMe updated
- Tests moved to other package to avoid dependency

batiati committed:
clients/java: Exceptions rework

- Created InitializationException and RequestException holding the status code

batiati committed:
clients/java: Java format

batiati committed:
clients/java: Remove unused imports

batiati committed:
clients/java: Exception for empty message body

batiati committed:
clients/java: Add asserts

Assert if the received buffer len is aligned with the data type size

Still needs to assert if the received operation matches the resquest operation flag, and if both request and reply has equals or less elements than the request

batiati committed:
clients/java: ReadMe

batiati committed:
clients/dotnet: Install scripts

batiati committed:
clients/dotnet: ReadMe

batiati committed:
clients/dotnet: ReadMe

batiati committed:
clients/java: Usage

- Usage ReadMe section
- Include new constructor with default maxConcurrency
- Include negative value checks for Transaction.setAmount()

batiati committed:
clients/java: Windows support

- Workarround a lld-link error building on windows
- Include .bat scripts
- Update ReadMe

batiati committed:
clients/dotnet: Windows scripts

- Added install.bat and benchmark.bat
- Updated ReadMe.md

batiati committed:
clients/java: Client.close

- Wait for processing threads to finish before deinitializing the tb_client

batiati committed:
clients/java: Java format

batiati committed:
clients/java: Asserts, exceptions and concurrency

- Client.close waits for running requests to finish
- New requests during and after the close will throw IllegalStatusException
- Added asserts on the Zig side
- Callback checks if the received data and operation matches the request

batiati committed:
clients/java: More tests

- TransfersBatchTest.java

batiati committed:
clients/dotnet: Synchronization tests

- Added proprer synchronization when disposing the client
- Waiting for an available packet
- Concurrency tests

batiati committed:
clients/java: Fix close sync

batiati committed:
clients/dotnet: Fix close sync

batiati committed:
clients/java: More tests

- Integrations tests with a live tb server

batiati committed:
clients/java: Initial concurrency testsy

eatonphil committed:
clients/node: pending field is now pending_id

eatonphil committed:
clients/node: Merge pull request #30 from tigerbeetledb/pe/pending-field

pending field is now pending_id

batiati committed:
clients/java: Refactory and code review

- JNI refactored following the JNI  best practices guide
https://developer.ibm.com/articles/j-jni/

- Fixed a bug when pop/pushing a packet from packetList

- Concurrency integration test passing

- Java semaphore now uses fairness=false, since it is not necessary to guarantee the order of the acquiring threads.

- Minor refactory on Java code, adding several assertions.

batiati committed:
clients/java: More tests

batiati committed:
clients/java: Update submodule

batiati committed:
clients/java: Benchmark moved

- Benchmark moved from Client.java to Benchmark.java outside package's source tree.

batiati committed:
clients/java: Minor refactory

- More assertions on the JNI side
- Sufix _obj for all jobject references

batiati committed:
clients/java: Unit tests

- All units tests have been moved to a separated source root
Reasoning:
(1) This practice allows testing protected classes and methods, even in separated roots, they share the same package name, thus the same visibility,
and
(2) allows easier separation between production code and test classes, avoiding complex build or unnecessary dependencies.

- Added unit tests for all protocol-related code
Even though the Tigerbeetle protocol does not require any serialization, we must assert the correct order of all fields written or read from the stream

- Removed non-ergonomic enums AccountFlags and TransactionFlags in favor of a simple integer constant.
Maybe implementing a helper to combine the flags would be a better solution

- Wrong byte order when reading/writing a UUID
Java is essentially bid-endian in its API, just swapped the parameters order on UUID's constructor.

batiati committed:
clients/java: Merge pull request #1 from batiati/master

Unit tests

eatonphil committed:
clients/node: Unswap code and ledger values in transfer example (#31)

* Unswap code and ledger values in transfer example

* Just swap values

batiati committed:
clients/java: Exceptions rework

- Better ergonomics when catching exceptions, making all checked exceptions we throw inherit from Exception class.

- Removed unnecessary "throws" block for unchecked exceptions

- New "AssertionError" is thrown on any unexpected condition detected during the execution.

This approach of throwing an "Error" instead of an "Exception", makes explicit to the user that this result is unexcepted and should only be logged and handled at a high level.

batiati committed:
clients/java: More exception rework

- InitializationException is now an unchecked exception

Rationale: It reduces a lot of friction by not requiring the user to explicitly handle an exception that cannot be recovered from.

Also, the INVALID_ADDRESS result is now handled as IllegalArgumentException, along with other argument validations, making it more idiomatic.

- Removed InterruptedException from the "throws" block.

Rationale: We don't provide ways for the user to cancel an ongoing request, so there is no point for them to explicitly handle this exception.

Actually, this exception is not expected to be thrown ever, the only way for it to happen is somehow by Thread.interrupt() being called. In this case, we abort throwing an AssertionError

- Minor code cleanup, removing unused imports.

batiati committed:
clients/java: More exceptions rework

- Preserves the Future<T> interface by not removing the InterruptedException from there.
Rationale: This is a valid use case of InterruptedException that was not covered by the previous commit.
Some async framework could use this exception to interrupt and resume a Future<T> anytime.
Our sync API now is using the "waitForCompletionUninterruptibly" and still assumes that no interruption will ever occur, not exposing it on the API.
Only when a request is used as Future<T> this exception is exposed as required by the interface.

batiati committed:
clients/java: AssertionError now inherits from java.lang.AssertionError

Rationale:
- We can filter logs and track this kind of error by our classname com.tigerbeetle.AssertionError
- Log frameworks and analysis tools can handle this error like any other generic java.lang.AssertionError

batiati committed:
clients/java: Merge pull request #2 from batiati/master

Exceptions rework

batiati committed:
clients/java: Maven project

- Created a pom.xml file
- Folder struct migrated to Maven-style
- JNILoader allowing to load a bundled native library inside the jar
- Basic build.zig file cross-compiling
- Basic goals for comiling, testing, javadocs and packaging
- Initial GHA manifest

Still needs work:

- Review and remove unecessary parts from the pom.xml file
- Define the minimum required JVM version and assert it during the build
- Configure the .vscode/settings.config file for Maven
- Review and assert if the generated package was correctly generated
- Adjust the build.zig file

batiati committed:
clients/java: Maven project

- Minor adjustments on install/benchmark scripts
- Fixed build.zig and documented the workarrounds

Still needs work:

- Address the issues on build.zig

batiati committed:
clients/java: Maven project

- General cleanup on pom.xml
- Fix the exec plugin order to build the JNI before the compilation
- Remove the JNI build step from install.sh, since it is done by Maven now
- Benchmark scripts now use mvn compile instead of javac

batiati committed:
clients/java: Maven build

- Updates the JUI submodule
- Updates the build.zig fixing a cross-compilation issue
- Adds more metadata on the pom.xml file, and bumps the version for testing purposes

batiati committed:
clients/java: GHA

- Assert the min required JVM version

batiati committed:
clients/java: GHA

- Asserts the min required JVM version

Defined to be Java 11 LTS, since Java 8 have reached its EOF already.
https://blogs.oracle.com/java/post/end-of-public-updates-is-a-process-not-an-event

batiati committed:
clients/java: Formatter

- formatter.xml based on the Google Java Style Guide
https://google.github.io/styleguide/javaguide.html
Minor changes regarding the use of 4 spaces for tabs

- Included a liter step on GHA
Just for the java code for now
Zig code still needs a little adjustment to fit into the TIGER_STYLE.MD (mostly due to the use of snake_case for functions)

- Removed old values from settings.json

batiati committed:
clients/java: First format

- Just applying the new formatting rules to the existing code

batiati committed:
clients/java: Format adjustments

- Minor adjustments on code formatting

batiati committed:
clients/java: Format adjustments

- More format adjustments on the test project

batiati committed:
clients/java: Windows support

- Tested on a Windows machine
Cross compilation to Windows still doesn't work

batiati committed:
clients/java: GHA fix

- Run the "Publish" step only when pushing

batiati committed:
clients/java: Use mvn in quiet mode in the benchmark

batiati committed:
clients/java: Make constants from TB's protocol more explicit

batiati committed:
clients/java: Formatting fix

Prevents unwanted line breaks by not using the fluent-style invocation.

sentientwaffle committed:
clients/go: Update URLs to new repo

batiati committed:
clients/java: Use the auto-generated GITHUB_TOKEN

batiati committed:
clients/java: Merge pull request #3 from batiati/master

Maven project and GHA

batiati committed:
clients/java: Rename repo owner

batiati committed:
clients/java: Readme update

- Repo links updated

- More emphasis on the usage section, explaining how to download the pre-built jar package

- Better code example snippets

- Deployment section includes dependency accordingly to the new Maven project

batiati committed:
clients/java: ReadMe fixes

- Missing period
- Incorrect variable name

batiati committed:
clients/java: Merge pull request #4 from batiati/readme

Readme update

batiati committed:
clients/java: Zig code review

- Function names to snake_case following TB's code style

- Minor refactory and reorganization

- Better comments

- Use of AssertionError instead of panic whenever the Java side can handle the error

batiati committed:
clients/java: Zig code review

- Refactored build.zig, using tb_client.zig as a zig package, instead of compiling as a static library.
- Cross-compiling to windows
- Minor refactory on client.zig, matching type names from tb_client
- Added a test block with comments justifying the lack of tests

batiati committed:
clients/java: Zig code review

- Panic rework

batiati committed:
clients/java: Bump junit version

batiati committed:
clients/java: GHA adjust

- Avoid executing tests twice when building and deploying

batiati committed:
clients/java: Merge pull request #6 from batiati/master

Updates JUnit dependency

batiati committed:
clients/java: Merge branch 'tigerbeetledb:master' into zig-review

batiati committed:
clients/java: Account

- Unit tests for all acessors
- Javadoc describing all acessors, parameters, return values and exceptions

batiati committed:
clients/java: More tests

- Transfer acessors tests
- Transfer Javadocs
- More coverage for existing tests
- Removed wrong "throws" block

batiati committed:
clients/java: Minor refactory

- Remove wrong "throws" block
- Rephrase comments

batiati committed:
clients/java: More tests

- Request unit tests
- Removed wrong "throws" block
- More javadoc

batiati committed:
clients/java: More unit tests

- Constants are now declared on interfaces, instead of "final classes"
It is more idiomatic

- More asserts and parameter validations

- More javadocs

batiati committed:
clients/java: More unit tests

- JNILoader tests
- More Javadocs

batiati committed:
clients/java: Minor adjustments

- Add a message to all NullPointerException
- Remove random values from tests

batiati committed:
clients/java: More tests

- Adds missing test for TOO_MUCH_DATA result
- Adds test for a null element on the replicaAddresses array

Javados

- Updates javadoc on NullPointerException regarding null arrays and null elements.

Fixes

- Fixes the assert order for "buffer == null", since it's expected to receive a null bullfer when "status != OK"

Minor rework

- Uses "import statis" instead of "Assert.assertXYZ" on IntegrationTests.java, following the other tests.
- Removes the "exception swallowing" on "IntegrationTests.Server.cleanUp"

batiati committed:
clients/java: Initial async rework

- Replaces Future<T> for CompletableFuture<T>
- Removes specifics Request classes for each kind of batch
- Creates BlockingRequest and AsyncRequest, reducing the complexity of a single implementation for both blocking and asynchronous waiting

batiati committed:
clients/java: Javadoc corrections

batiati committed:
clients/java: Javadoc corrections

batiati committed:
clients/java: Adds test coverage thresholds

- GHA will fail with less than 95% of tested intructions or less than 90% of tested branches.
We have low branch ratio due to asserts on unexpected values that are unreachable.

- The current values on this commit are 96% and 93% respectively.

batiati committed:
clients/java: Adds test coverage thresholds

- GHA will fail with less than 95% of tested intructions or less than 90% of tested branches.
We have low branch ratio due to asserts on unexpected values that are unreachable.

- The current values on this commit are 96% and 93% respectively.

batiati committed:
clients/java: Merge pull request #7 from batiati/unit-tests

Merging unit tests and initial java documents.
API changes will be addressed in an upcoming PR.

batiati committed:
clients/java: Merge branch 'api-rework' into async-rework

batiati committed:
clients/java: Merge pull request #1 from batiati/async-rework

Async rework

batiati committed:
clients/java: Initial API rework

- Removed UUID from the API
- UInt128 exposed as a pair of long or as a byte array

batiati committed:
clients/java: More API rework

- Batch implements a "Cursor" interface, allowing a single instance to point to multiple elements, depending on the current position.
There is an specialized batch for each struct: Accounts, Transfers, etc.

- Removed Account, Transfers, Id, CreateTransferResults and CreateAccountResults
These structures are no longer exposed individually and can only be accessed from one of the specialized batches.

- Client API uses the new Batch

- Some tests were rewritten

(...)

Still needs work

- Complete all tests

- Define better names for some classes/methods
ex: Batch.add(), Batch.reset(), UInt128.Bytes, etc

batiati committed:
clients/java: Formatter

batiati committed:
clients/java: More tests

- More Batch tests passing for Accounts and Transfers

- Fixed a false-negative test for closing a client in concurrent scenarios

- Introduced a new method AssertionError.assertTrue
Using a method instead of a IF removes the missed-branches metric for assertions

- Minor code cleanup

batiati committed:
clients/java: Update tigerbeetle subrepo

batiati committed:
clients/java: More API rework + tests finished

- Batch.add now is protected
Only writable batches (Ids, Accounts, Transfers) can expose this method
CreateAccountResults, CreateTransferResults expose a read-only API.

- UInt128 abstraction
It works as an static helper class, converting from/to long, byte[] and UUID

- More tests, increasing the coverage to 97%

Pending work

- Add javadocs to the new API

- Review throws on the new API

batiati committed:
clients/java: Javadoc updates

batiati committed:
clients/java: Fix javadoc

I don't know what I was thinking when I wrote "transaction" instead of
"transfer" ;-)

batiati committed:
clients/java: More javadocs

batiati committed:
clients/java: Formatter

batiati committed:
clients/java: Fix intermittent test failure

- testCloseWithConcurrentTasks may fail
Client.close tried to acquire atomically all permits before closing,
this sometimes caused all thread to succeed, while the test expected
at least one failure.

Changed for a loop acquiring each permit individually.

batiati committed:
clients/java: Updates README.md

batiati committed:
clients/java: Updates README.mc

batiati committed:
clients/java: Spellchecker 😅 🤓

batiati committed:
clients/java: Review

- Renamed all Batch sub-classes

- Link to the official docs

- Minor adjustments

batiati committed:
clients/java: README.md

batiati committed:
clients/java: Merge pull request #8 from batiati/api-rework

Api rework

batiati committed:
clients/java: Update README.md

- Fix code examples.

batiati committed:
clients/java: Module info

- Added the module name "com.tigerbeetle"

- Added module dependencies for tests

- Removed unnecessary reference to javax.management

- Added mandatory package name to the Benchmark program.
Note: The benchmark program is still part of the main package and must be moved in a next PR.

batiati committed:
clients/java: Merge pull request #12 from batiati/modules

Closes #11

batiati committed:
clients/java: Merge pull request #3 from batiati/zig-review

Merge from PR #5

batiati committed:
clients/java: Fix cross-compiling from Windows and MacOs

batiati committed:
clients/java: Setup PMD plugin

- Plugin for static analysis and coding style
https://pmd.github.io
Changes applied and the exclusions have been detailed in the exclude-pmd.properties file.

batiati committed:
clients/java: Initial demo program

- Project structure

- README + descriptions

- Basic StartHere.java demo

batiati committed:
clients/java: Demo program

batiati committed:
clients/java: Demo program

batiati committed:
clients/java: Demo program

sentientwaffle committed:
clients/node: Update TigerBeetle: Add `linked_event_chain_open`

batiati committed:
clients/java: Demo program

- UInt128 helpers for BigInteger

batiati committed:
clients/java: Update TigerBeetle: Add linked_event_chain_open

- TigerBeetle updated, see tigerbeetledb/tigerbeetle#176

- Mocked test results updated with new error codes.

- Bumped the number of transfers on integration tests, as the server goes more mature.

batiati committed:
clients/java: Formatter

batiati committed:
clients/java: Merge pull request #13 from batiati/rebase

Update TigerBeetle: Add linked_event_chain_open

batiati committed:
clients/java: Rebase

batiati committed:
clients/java: Passing tests

batiati committed:
clients/java: Pom exec plugin

batiati committed:
clients/java: Project exec and command line args

batiati committed:
clients/java: Demo program

batiati committed:
clients/java: README

batiati committed:
clients/java: Demo program

batiati committed:
clients/java: Demo program

batiati committed:
clients/java: Scripts for running a stand-alone replica

batiati committed:
clients/java: Windows support

batiati committed:
clients/java: Examples

- Rename "Samples" and "Demo" to "Examples"

batiati committed:
clients/java: Examples

- Rename "samples" and "demo" to "examples"

batiati committed:
clients/java: Script

batiati committed:
clients/java: Screenshots update

batiati committed:
clients/java: Batch.next end of batch assertion

- Calling Batch.next multiple times must throw an exception.
This change prevents the user from assuming that an iterated batch is an empty one.
Also it follows the java.sql.ResultSet.next() behavior.

- Introduced an assertion if we are running on a LITTLE ENDIAN platform.
The C TBClient would assert that anyway.

batiati committed:
clients/java: Update tb_client

See https://github.com/tigerbeetledb/tigerbeetle/pull/182

batiati committed:
clients/java: Merge pull request #15 from batiati/rebase

Update tb_client

sentientwaffle committed:
clients/node: Update TigerBeetle: Forbid new accounts with amount≠0

batiati committed:
clients/java: Forbid new accounts with amount ≠ 0

See https://github.com/tigerbeetledb/tigerbeetle/pull/192

batiati committed:
clients/java: Forbid new accounts with amount ≠ 0

See https://github.com/tigerbeetledb/tigerbeetle/pull/192

batiati committed:
clients/dotnet: Update tigerbeetle

batiati committed:
clients/dotnet: Code cleanup

batiati committed:
clients/dotnet: README

batiati committed:
clients/java: Rollback examples

- For simplicity, the examples will be part of another PR

batiati committed:
clients/java: Minor adjustments

- POM linkXRef=false removing the PMD warning about the undefined jxr reporting plugin configuration.
Since we are not using the PDM reporting there is no need to configure it.

- Using mvn --quiet mode on the install scripts

batiati committed:
clients/java: Update tigerbeetle

batiati committed:
clients/java: Reverting submodule update

batiati committed:
clients/java: DX improvements

batiati committed:
clients/java: Example program

- Added 3 example programs: QuickStart.java, LinkedEvents.java and PendingTransfers.java

- README
Instructions on how to build and run the examples
Brief description with screenshots.

eatonphil committed:
clients/node: Add install tests on various common distros (#35)

* Start distro install tests

* Add dedicated build script via docker for glibc alignment

* Fedora tests are working

* Add checkout step

* Rename references to coilhq

* Update tigerbeetle

* Add shellcheck

* Bump submodule

* Shellcheck and alpine tests working

* Test amazonlinux

* RHEL and AmazonLinux are working, drop RockyLinux for now

* sh for alpine

* Add success note

* Checkout submodules

* Add missing build step for alpine

* Update readme

eatonphil committed:
clients/node: Add error handling info to docs, mix less source code and client code, links to docs.tigerbeetle.com (#40)

* Add transfer lookup

* Add note about query api

* Update README.md

Co-authored-by: djg <sentientwaffle@gmail.com>

* Update README.md

Co-authored-by: djg <sentientwaffle@gmail.com>

* Update README.md

Co-authored-by: djg <sentientwaffle@gmail.com>

* Drop tabs

* Fix switch syntax

* Break switch

* Take DJs suggestion

Co-authored-by: djg <sentientwaffle@gmail.com>

eatonphil committed:
clients/node: Fix TransferFlags (#41)

sentientwaffle committed:
clients/go: Update: linked_event_chain_open

Also account balances are now initially zero.

kprotty committed:
clients/go: Regenerate tb_client Go types

Commit:
clients/go: Regenerate tb_client header + types

kprotty committed:
clients/go: Re-sync with tigerbeetle main

sentientwaffle committed:
clients/node: Update TigerBeetle: remove a couple error codes (#42)

batiati committed:
clients/java: Update submodule

batiati committed:
clients/java: CreateAccountResult enum changes

- Removed results for overflows/exceeds credits and debits
See https://github.com/tigerbeetledb/tigerbeetle/pull/141 for more details

batiati committed:
clients/java: Batch bugfix

Fixes a bug introduced at
https://github.com/tigerbeetledb/tigerbeetle-java/commit/096e7010c3ab094f422171f704993213578bf1bf

Empty batches share a single instance, avoiding allocations for the common case (empty results == success),
and calling next() em empty results would fail instead of returning false.

This fix prevents an empty batch from moving the cursor ahead, keeping the validation introduced for iterated batches.

Tests were updated to reflect this change.

batiati committed:
clients/java: Merge pull request #18

Updates and fixes

batiati committed:
clients/java: Merge branch 'tigerbeetledb:master' into examples

batiati committed:
clients/java: Example programs CI validation

jorangreef committed:
clients/node: 0.11.0

batiati committed:
clients/dotnet: Update tb_client

eatonphil committed:
clients/go: Add sample program and Github CI steps, add helper function for getting uint128s (#15)

Add shellcheck, rebuild_binaries script, and README for basic sample

batiati committed:
clients/dotnet: Update binaries

batiati committed:
clients/java: Link to the examples page

batiati committed:
clients/java: Example program

batiati committed:
clients/java: Update C client

- Updates tigerbeetle submodule
- Adds tests covering the issue https://github.com/tigerbeetledb/tigerbeetle/issues/230
- Improves the client log

batiati committed:
clients/java: Update C client

batiati committed:
clients/java: C client update

- Updates the submodule path (still had coilhq reference in the name).

- Minor changes due to https://github.com/tigerbeetledb/tigerbeetle/pull/237 adjusting `tb_packet_t` field types to reflect C header.

batiati committed:
clients/java: C client update #20

cb22 committed:
Merge remote-tracking branch 'tigerbeetle-node/main'

cb22 committed:
Merge remote-tracking branch 'tigerbeetle-dotnet/main'

cb22 committed:
Merge remote-tracking branch 'tigerbeetle-java/master'

cb22 committed:
clients: Import clients and scope CI workflows into main repo

cb22 committed:
Merge remote-tracking branch 'tigerbeetle-go/main'

sentientwaffle commented on .github/workflows/clients-go-build.yml:
```diff
@@ -0,0 +1,136 @@
+          restore-keys: |
+            go-binaries-${{ github.sha }}
+      - run: mkdir ~/.ssh && echo -e "${{ secrets.TIGERBEETLE_GO_DEPLOY_KEY }}" > ~/.ssh/id_ed25519 && chmod 600 ~/.ssh/id_ed25519
+      - run: git clone git@github.com:tigerbeetledb/tigerbeetle-go.git ~/tigerbeetle-go && cp -R . ~/tigerbeetle-go
```
If a file is moved in `tigerbeetledb/tigerbeetle`, will this leave the old copy intact in `tigerbeetledb/tigerbeetle-go`? Maybe we need to delete the files (except `.git`) in `~/tigerbeetle-go` before copying `./src/clients/go/*` in.

cb22 replied:
Good catch - we need to do exactly that, will sort it out.

sentientwaffle commented on .github/workflows/clients-node-build.yml:
```diff
@@ -0,0 +1,52 @@
+      - run: npm install && ./scripts/version_check.sh
+
+  publish:
+    if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/main' }}
```
Just to check my understanding, to publish a new version, we just need to:

- bump the `version` in `package.json`,
- and then merge the commit that bumps the version into `main`.

cb22 replied:
Yep!

sentientwaffle replied:
That's perfect; thank you!

sentientwaffle commented on src/clients/node/package.json:
```diff
@@ -39,7 +40,7 @@
     "build_tsc": "./node_modules/typescript/bin/tsc",
     "build_lib": "mkdir -p dist && zig/zig build-lib -mcpu=baseline -OReleaseSafe -dynamic -lc -isystem build/node-$(node --version)/include/node src/node.zig -fallow-shlib-undefined -femit-bin=dist/client.node",
-    "prepack": "git submodule deinit --all && git submodule update --init && npm run build_tsc",
+    "prepack": "npm run build_tsc && rm -rf src/zig-cache",
```
Why is `rm -rf src/zig-cache` needed? My understanding of `npm` (maybe wrong) is that the `files` list (above) enumerates the files packaged.

cb22 replied:
It's not anymore, I'll take that out. I originally added it before I modified the `files` to be more specific.

sentientwaffle reviewed (commented):
The deploy & publish keys are set up.

eatonphil commented:
Nice work, Fed!

eatonphil commented on .github/workflows/clients-go-build.yml:
```diff
@@ -0,0 +1,136 @@
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - run: ./scripts/shellcheck.sh
```
We can probably now drop all individual calls to shellcheck and just have one global one?

cb22 replied:
Yep! I can do this in here, or as a separate commit once this is merged?

eatonphil replied:
Whatever you'd like

sentientwaffle replied:
Either is fine; whichever is simpler for you.

eatonphil commented on .github/workflows/clients-go-build.yml:
```diff
@@ -0,0 +1,136 @@
+            go-binaries-${{ github.sha }}
+      - run: go test ./...
+
+  # TODO: Windows exit codes broken
```
```
Set-StrictMode -Version 3
$ErrorActionPreference = "Stop"
$PSDefaultParameterValues['*:ErrorAction']='Stop'
$LASTEXITCODE = 0
```

Here's what I do [elsewhere](https://github.com/multiprocessio/datastation/blob/main/scripts/ci/prepare_windows.ps1).

cb22 replied:
Cool thanks - I'll get this sorted

eatonphil commented on .github/workflows/clients-node-build.yml:
```diff
@@ -0,0 +1,52 @@
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - run: ./scripts/shellcheck.sh
```
This could be dropped since a global check should catch this

eatonphil commented:
> To keep binaries out of the main repo, the Go client workflow builds the binaries needed and then pushes the result to the tigerbeetle-go repo. It includes the source commit in its commit message, as well as creates a tag called tigerbeetle-{commit-in-this-repo} for easy reference.

I don't understand the benefit of keeping binaries out of the repo.

sentientwaffle commented:
> I don't understand the benefit of keeping binaries out of the repo.

This was my request — I prefer not to include them b/c they make checkouts larger, and my workflow involves around half a dozen branches checked out simultaneously (via git worktrees) :wink:.

eatonphil commented:
> This was my request — I prefer not to include them b/c they make checkouts larger, and my workflow involves around half a dozen branches checked out simultaneously (via git worktrees) 😉.

Gotcha.

cb22 committed:
clients: Use repository secrets, clear the tigerbeetle-go directory, remove unneeded rm

cb22 commented:
I've pushed a commit addressing some of the issues above - the Go repo files & npm prepack.

The `.bat` Windows tests and shell-check I'll do in a separate PR, just to keep things more manageable!

We'll also need to enable some more actions in our project settings:

- `actions/checkout@v3`
- `actions/cache@v3`
- `actions/setup-java@v3`
- (or perhaps `actions/*`?)

sentientwaffle commented:
I added
```
actions/checkout@v3
actions/cache@v3
actions/setup-java@v3
```
to our allowed-actions list.
